### PR TITLE
Create a new instance of aws cloud

### PR DESCRIPTION
**Pull request by nicolehanjing**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This PR creates a new instance of aws cloud, which helps future PRs to commit in parallel with fewer conflicts

**Which issue(s) this PR fixes**:

Part of #125

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Create a new instance of aws cloud of v2 provider.
```

### CODE REVIEW

File: pkg/providers/v2/interface.go [hunk 1]
```diff
@@ -41,6 +48,86 @@ var _ cloudprovider.Interface = (*cloud)(nil)
 type cloud struct {
 }
 
+type awsSDKProvider struct {
+	creds *credentials.Credentials
```
Comment (andrewsykim) [inline pkg/providers/v2/interface.go:52]: Not sure we need a new type like `awsSDKProvider` yet, seems premature.  Can we have `creds` and the region field be in `cloud` instead?

File: pkg/providers/v2/interface.go [hunk 2]
```diff
@@ -41,6 +48,86 @@ var _ cloudprovider.Interface = (*cloud)(nil)
...
+	GetMetadata(path string) (string, error)
+}
+
+func getAvailabilityZone(metadata EC2Metadata) (string, error) {
+	return metadata.GetMetadata("placement/availability-zone")
+}
+
+// Derives the region from a valid az name.
+// Returns an error if the az is known invalid (empty)
+func azToRegion(az string) (string, error) {
```
Comment (andrewsykim) [inline pkg/providers/v2/interface.go:67]: Can we add a unit test for this?

Comment (HaibaraAi96) [reply pkg/providers/v2/interface.go:67]: sure, will add soon!

File: pkg/providers/v2/interface.go (hunk 1 above)
Comment (HaibaraAi96) [reply pkg/providers/v2/interface.go:52]: that sounds reasonable, will update!

File: pkg/providers/v2/interface.go (hunk 2 above)
Comment (HaibaraAi96) [reply pkg/providers/v2/interface.go:65]: @andrewsykim Updated!
I add a new file called interface_test.go and include the unit test for it
I also delete the type awsSDKProvider and call metadataClient directly in newCloud()

File: pkg/providers/v2/interface.go [hunk 3]
```diff
@@ -39,6 +46,71 @@ var _ cloudprovider.Interface = (*cloud)(nil)
 
 // cloud is the AWS v2 implementation of the cloud provider interface
 type cloud struct {
```
Comment (andrewsykim) [inline pkg/providers/v2/interface.go:48]: Line 36 above needs to be updated to call newCloud as part of `RegisterCloudProvider`

File: pkg/providers/v2/interface.go [hunk 4]
```diff
@@ -39,6 +46,71 @@ var _ cloudprovider.Interface = (*cloud)(nil)
...
+	region := r.FindString(az)
+	if region == "" {
+		return "", fmt.Errorf("invalid AZ: %s", az)
+	}
+
+	return region, nil
+}
+
+// newCloud creates a new instance of AWSCloud.
+func newCloud() (*cloud, error) {
```
Comment (andrewsykim) [inline pkg/providers/v2/interface.go:80]: This should return the interface it implements instead
```
func newCloud() (cloudprovider.Interface, error)
```

**Review by andrewsykim — COMMENTED** [review]
/ok-to-test

File: pkg/providers/v2/cloud.go [hunk 5]
```diff
@@ -39,6 +46,71 @@ var _ cloudprovider.Interface = (*cloud)(nil)
...
+}
+
+func getAvailabilityZone(metadata EC2Metadata) (string, error) {
+	return metadata.GetMetadata("placement/availability-zone")
+}
+
+// Derives the region from a valid az name.
+// Returns an error if the az is known invalid (empty)
+func azToRegion(az string) (string, error) {
+	if len(az) < 1 {
```
Comment (andrewsykim) [inline pkg/providers/v2/cloud.go:66]: nit: this probably reads slightly better as:
```
if len(az) == 0
```

Comment (HaibaraAi96) [reply pkg/providers/v2/cloud.go]: gotcha, updated!

File: pkg/providers/v2/cloud.go [hunk 6]
```diff
@@ -19,14 +19,21 @@ limitations under the License.
...
+	"github.com/aws/aws-sdk-go/aws/credentials/ec2rolecreds"
+	"github.com/aws/aws-sdk-go/aws/ec2metadata"
+	"github.com/aws/aws-sdk-go/aws/session"
 	cloudprovider "k8s.io/cloud-provider"
 )
 
 func init() {
 	cloudprovider.RegisterCloudProvider(ProviderName, func(config io.Reader) (cloudprovider.Interface, error) {
-		return &cloud{}, nil
+		return newCloud()
```
Comment (HaibaraAi96) [inline pkg/providers/v2/cloud.go:36]: @andrewsykim updated! 
Also passed the unit test, let me know if you have any other suggestions :)

**Review by andrewsykim — COMMENTED** [review]
/approve
/lgtm

thanks @nicolehanjing

**Pull request merged**